File: Algorithms/Medium/DesignBrowserHistory.go
package medium

import "github.com/parviz-yu/LeetCode/Algorithms/helpers"

type BrowserHistory struct {
	history []string
	ptr     int
	len     int
}

func NewBrowserHistory(homepage string) BrowserHistory {
	return BrowserHistory{
		history: []string{homepage},
		ptr:     0,
		len:     1,
	}
}

// Time Complexity: O(1)
func (this *BrowserHistory) Visit(url string) {
	if len(this.history) < this.ptr+2 {
		this.history = append(this.history, url)
	} else {
		this.history[this.ptr+1] = url
	}
	this.ptr++
	this.len = this.ptr + 1
}

// Time Complexity: O(1)
func (this *BrowserHistory) Back(steps int) string {
	this.ptr = helpers.Max(0, this.ptr-steps)
	return this.history[this.ptr]
}

// Time Complexity: O(1)
func (this *BrowserHistory) Forward(steps int) string {
	this.ptr = helpers.Min(this.len-1, this.ptr+steps)
	return this.history[this.ptr]
}

/**
 * Your BrowserHistory object will be instantiated and called as such:
 * obj := Constructor(homepage);
 * obj.Visit(url);
 * param_2 := obj.Back(steps);
 * param_3 := obj.Forward(steps);
 */

File: Algorithms/Medium/1472.Design-Browser-History/DesignBrowserHistory.py
class Node:
    def __init__(self, value: str, next=None, prev=None):
        self.value = value
        self.next = next
        self.prev = prev


class BrowserHistory:
    def __init__(self, homepage: str):
        self.curr_page = Node(homepage)

    # Time complexity: O(1)
    def visit(self, url: str) -> None:
        new_page = Node(url, prev=self.curr_page)
        self.curr_page.next = new_page
        self.curr_page = new_page

    # Time complexity: O(n)
    def back(self, steps: int) -> str:
        while self.curr_page.prev and steps > 0:
            self.curr_page = self.curr_page.prev
            steps -= 1

        return self.curr_page.value

    # Time complexity: O(n)
    def forward(self, steps: int) -> str:
        while self.curr_page.next and steps > 0:
            self.curr_page = self.curr_page.next
            steps -= 1

        return self.curr_page.value
